### PR TITLE
[FIX] Prevent cst duplicate imports

### PR DESCRIPTION
**Pull request by mohammedahmed18**
This fixes the issue here: [https://github.com/misrasaurabh1/recce/pull/30/commits/16f0638dc3840c8c1309fba8389b0421c00cdd31](https://github.com/misrasaurabh1/recce/pull/30/commits/16f0638dc3840c8c1309fba8389b0421c00cdd31)

Basically, `from recce.adapter.base import BaseAdapter` got added even though it was already in the file.

my guess is that `AddImportsVisitor.add_needed_import` stops looking for imports once it hits something that’s not an import. In this case, it hit the `try/except` block within lines 39:46, so it never saw the existing `BaseAdapter` import at line 50 and just added it again.

The fix was to manually collect all top-level imports in a normalized dotted format, so duplicates don’t slip through anymore.
___

### **Description**
- Replace conditional import collector with dotted importer

- Prevent adding duplicate imports in extraction logic

- Update import normalization and alias handling

- Add test for duplicate import removal

___

### Diagram Walkthrough

```mermaid
flowchart LR
  parse["Parse destination module"] --> collect["DottedImportCollector.collect imports"]
  gather["Gather needed imports"] --> filter["Filter out existing imports"]
  filter --> add["AddImportsVisitor.add_needed_import"]
  filter --> remove["RemoveImportsVisitor.remove_unused_import"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>code_extractor.py</strong><dd><code>Refactor import collector and dedupe logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

codeflash/code_utils/code_extractor.py

<ul><li>Rename ConditionalImportCollector to DottedImportCollector<br> <li> Normalize top-level imports to dotted format<br> <li> Replace collector usage and invert duplicate import checks<br> <li> Invoke block import collection in visit_Module</ul>

</details>

  </td>
  <td><a href="https://github.com/codeflash-ai/codeflash/pull/665/files#diff-d76cd127e1798f120f04726e1a3775840a9870e224395fd25bae32235c6982d0">+31/-17</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_add_needed_imports_from_module.py</strong><dd><code>Add test for duplicate import removal</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_add_needed_imports_from_module.py

<ul><li>Add import <code>re</code> for regex assertions<br> <li> Introduce <code>test_duplicated_imports</code> for dedupe check<br> <li> Assert only one BaseAdapter import remains</ul>

</details>

  </td>
  <td><a href="https://github.com/codeflash-ai/codeflash/pull/665/files#diff-761e9ed15c22dc26c56bc95df70d3267258bb83485ac6d76a8f1a0df0ac19c8f">+127/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: tests/test_add_needed_imports_from_module.py [hunk 1]
```diff
@@ -121,3 +122,129 @@ def belongs_to_function(name: Name, function_name: str) -> bool:
...
+
+        return parent_map
+'''
+    src_path = Path("/home/roger/repos/codeflash/cli/codeflash/optimization/function_context.py")
+    dst_path = Path("/home/roger/repos/codeflash/cli/codeflash/optimization/function_context.py")
+    project_root = Path("/home/roger/repos/codeflash")
+    new_module = add_needed_imports_from_module(src_module, dst_module, src_path, dst_path, project_root)
+
+    matches = re.findall(r"^\s*from\s+recce\.adapter\.base\s+import\s+BaseAdapter\s*$", new_module, re.MULTILINE)
+    assert len(matches) == 1, f"Expected 1 match for BaseAdapter import, but found {len(matches)}"
```
Comment (misrasaurabh1) [inline tests/test_add_needed_imports_from_module.py]: i would prefer a direct string comparison over the entirety of the newly replaced string. this way we have more of a guarantee than searching for particular patterns.

**Review by misrasaurabh1 — DISMISSED** [review]

**Review by misrasaurabh1 — APPROVED** [review]
